src/main/java/org/tikv/common/util/ChannelFactory.java

```diff
@@ -212,72 +319,122 @@ public ChannelFactory(
     this.keepaliveTime = keepaliveTime;
     this.keepaliveTimeout = keepaliveTimeout;
     this.idleTimeout = idleTimeout;
-    this.certContext = new JksContext(jksKeyPath, jksKeyPassword, jksTrustPath, jksTrustPassword);
+    this.connRecycleTime = connRecycleTime;
+    this.certContext =
+        new JksContext(path, jksKeyPath, jksKeyPassword, jksTrustPath, jksTrustPassword);
   }
 
-  @VisibleForTesting
-  public boolean reloadSslContext() {
-    if (certContext != null) {
-      SslContextBuilder newBuilder = certContext.reload();
-      if (newBuilder != null) {
-        sslContextBuilder.set(newBuilder);
-        return true;
+  private ManagedChannel createChannel(
+      SslContextBuilder sslContextBuilder, String address, HostMapping mapping) {
+    URI uri, mapped;
+    try {
+      uri = PDUtils.addrToUri(address);
+    } catch (Exception e) {
+      throw new IllegalArgumentException("failed to form address " + address, e);
+    }
+    try {
+      mapped = mapping.getMappedURI(uri);
+    } catch (Exception e) {
+      throw new IllegalArgumentException("failed to get mapped address " + uri, e);
+    }
+
+    // Channel should be lazy without actual connection until first call
+    // So a coarse grain lock is ok here
+    NettyChannelBuilder builder =
+        NettyChannelBuilder.forAddress(mapped.getHost(), mapped.getPort())
+            .maxInboundMessageSize(maxFrameSize)
+            .keepAliveTime(keepaliveTime, TimeUnit.SECONDS)
+            .keepAliveTimeout(keepaliveTimeout, TimeUnit.SECONDS)
+            .keepAliveWithoutCalls(true)
+            .idleTimeout(idleTimeout, TimeUnit.SECONDS);
+
+    if (sslContextBuilder == null) {
+      return builder.usePlaintext().build();
+    } else {
+      SslContext sslContext;
+      try {
+        sslContext = sslContextBuilder.build();
+      } catch (SSLException e) {
+        logger.error("create ssl context failed!", e);
+        throw new IllegalArgumentException(e);
       }
+      return builder.sslContext(sslContext).build();
     }
-    return false;
   }
 
-  public ManagedChannel getChannel(String addressStr, HostMapping hostMapping) {
-    if (reloadSslContext()) {
-      logger.info("invalidate connection pool");
-      connPool.clear();
+  private void cleanExpireConn(Collection<ManagedChannel> pending) {
+    for (ManagedChannel channel : pending) {
+      logger.info("cleaning expire channels");
+      channel.shutdownNow();
```

Comment:
Why use shutdownNow()? Maybe some channel is being used right now

Reply:
If the channel is used right now, no matter calling shutdown or `shutdownNow`, the user of the channel will receive an exception, a retry is expected.

Reply:
got it